File: app-presentation/src/main/java/com/zxjdev/smile/presentation/moment/list/MomentListPresenter.java
package com.zxjdev.smile.presentation.moment.list;

import android.os.Bundle;

import com.zxjdev.smile.domain.moment.Moment;
import com.zxjdev.smile.domain.moment.usecase.GetMomentList;
import com.zxjdev.smile.presentation.common.DefaultSubscriber;
import com.zxjdev.smile.presentation.common.util.ui.ErrorMessagePrinter;
import com.zxjdev.smile.presentation.moment.MomentModel;
import com.zxjdev.smile.presentation.moment.MomentModelMapper;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class MomentListPresenter implements MomentListContract.Presenter {

  private static final String EXTRA_MOMENT_LIST = "extra_moment_list";

  @Inject MomentListContract.View view;
  @Inject GetMomentList getMomentList;
  @Inject MomentModelMapper momentModelMapper;
  @Inject ErrorMessagePrinter errorMessagePrinter;

  private ArrayList<MomentModel> moments = new ArrayList<>();

  @Inject
  public MomentListPresenter() {

  }

  @Override
  public void create() {
    loadMoments();
  }

  @Override
  public void loadMoments() {
    getMomentList.execute(new DefaultSubscriber<List<Moment>>(errorMessagePrinter) {
      @Override
      public void onNext(List<Moment> data) {
        moments.clear();
        moments.addAll(momentModelMapper.transform(data));
        view.displayMomentList(moments);
      }
    });
  }

  @Override
  public void refreshMoments() {
    getMomentList.execute(new DefaultSubscriber<List<Moment>>(errorMessagePrinter) {
      @Override
      public void onNext(List<Moment> data) {
        moments.clear();
        moments.addAll(momentModelMapper.transform(data));
        view.displayMomentList(moments);
        view.dismissRefreshingView();
      }
    });
  }

  @Override
  public void saveInstanceState(Bundle outState) {
    outState.putParcelableArrayList(EXTRA_MOMENT_LIST, moments);
  }

  @Override
  public void loadSavedInstanceState(Bundle savedInstanceState) {
    if (savedInstanceState == null) return;

    ArrayList<MomentModel> moments = savedInstanceState.getParcelableArrayList(EXTRA_MOMENT_LIST);
    if (moments != null) {
      this.moments = moments;
      view.displayMomentList(moments);
    }
  }
}

File: app-presentation/src/main/java/com/zxjdev/smile/presentation/moment/list/MomentListPresenter.kt
package com.zxjdev.smile.presentation.moment.list

import android.arch.lifecycle.LifecycleOwner
import com.zxjdev.smile.domain.moment.Moment
import com.zxjdev.smile.domain.moment.usecase.GetMomentList
import com.zxjdev.smile.presentation.common.util.ui.ErrorMessagePrinter
import com.zxjdev.smile.presentation.moment.MomentModelMapper
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import javax.inject.Inject

class MomentListPresenter @Inject internal constructor() : MomentListContract.Presenter {

    private lateinit var view: MomentListContract.View
    @Inject lateinit var getMomentList: GetMomentList
    @Inject lateinit var momentModelMapper: MomentModelMapper
    @Inject lateinit var errorMessagePrinter: ErrorMessagePrinter
    @Inject lateinit var viewModel: MomentListViewModel
    @Inject lateinit var lifecycleOwner: LifecycleOwner
    private val compositeDisposable = CompositeDisposable()

    private fun loadMoments() {
        compositeDisposable.add(getMomentList.execute().subscribeWith(object : DisposableObserver<List<Moment>>() {
            override fun onComplete() {
            }

            override fun onNext(t: List<Moment>) {
                viewModel.setMoments(momentModelMapper.transform(t))
            }

            override fun onError(e: Throwable) {
                errorMessagePrinter.print(e.message)
            }
        }))
    }

    override fun refreshMoments() {
        compositeDisposable.add(getMomentList.execute().subscribeWith(object : DisposableObserver<List<Moment>>() {
            override fun onComplete() {
            }

            override fun onNext(t: List<Moment>) {
                view.dismissRefreshingView()
                viewModel.setMoments(momentModelMapper.transform(t))
            }

            override fun onError(e: Throwable) {
                errorMessagePrinter.print(e.message)
            }
        }))
    }

    override fun takeView(view: MomentListContract.View) {
        this.view = view
        if (viewModel.getMoments().value == null) {
            loadMoments()
        }
        viewModel.getMoments().observe(lifecycleOwner, android.arch.lifecycle.Observer { models ->
            view.displayMomentList(models)
        })
    }

    override fun dropView() {
        compositeDisposable.dispose()
    }
}
